rust/parser/literal.rs

```diff
@@ -0,0 +1,118 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::{
+    common::{error::TypeQLError, Spanned},
+    parser::{IntoChildNodes, Node, Rule, RuleMatcher},
+    value::{
+        BooleanLiteral, DateFragment, DateLiteral, DateTimeLiteral, DateTimeTZLiteral, IntegerLiteral, Literal, Sign,
+        SignedDecimalLiteral, SignedIntegerLiteral, StringLiteral, TimeFragment, TimeZone, ValueLiteral,
+    },
+};
+
+pub(super) fn visit_value_literal(node: Node<'_>) -> Literal {
+    debug_assert_eq!(node.as_rule(), Rule::value_literal);
+    let span = node.span();
+    let child = node.into_child();
+    let value_literal = match child.as_rule() {
+        Rule::quoted_string_literal => ValueLiteral::String(visit_quoted_string_literal(child)),
+        Rule::boolean_literal => ValueLiteral::Boolean(BooleanLiteral { value: child.as_str().to_owned() }),
+        Rule::signed_integer => ValueLiteral::Integer(visit_signed_integer(child)),
+        Rule::signed_decimal => ValueLiteral::Decimal(visit_signed_decimal(child)),
+
+        Rule::datetime_tz_literal => ValueLiteral::DateTimeTz(visit_datetime_tz_literal(child)),
+        Rule::datetime_literal => ValueLiteral::DateTime(visit_datetime_literal(child)),
+        Rule::date_literal => ValueLiteral::Date(visit_date_literal(child)),
+
+        _ => unreachable!("{}", TypeQLError::IllegalGrammar { input: child.to_string() }),
+    };
+    Literal::new(span, value_literal)
+}
+
+fn visit_sign(node: Node<'_>) -> Sign {
+    debug_assert_eq!(node.as_rule(), Rule::sign);
+    match node.as_str() {
+        "+" => Sign::Plus,
+        "-" => Sign::Minus,
+        _ => unreachable!("{}", TypeQLError::IllegalGrammar { input: node.to_string() }),
+    }
+}
+
+pub(super) fn visit_integer_literal(node: Node<'_>) -> IntegerLiteral {
+    debug_assert_eq!(node.as_rule(), Rule::integer_literal);
+    IntegerLiteral { value: node.as_str().to_owned() }
+}
+
+pub(super) fn visit_quoted_string_literal(node: Node<'_>) -> StringLiteral {
+    debug_assert_eq!(node.as_rule(), Rule::quoted_string_literal);
+    StringLiteral { value: node.as_str().to_owned() }
+}
+
+fn visit_signed_integer(node: Node<'_>) -> SignedIntegerLiteral {
+    debug_assert_eq!(node.as_rule(), Rule::signed_integer);
+    let mut children = node.into_children().collect::<Vec<_>>();
+    let integral = children.pop().unwrap();
+    let sign = children.pop().map(|node| visit_sign(node));
+    debug_assert_eq!(integral.as_rule(), Rule::integer_literal);
```

Comment:
Take a look at other visitors, `children` can and should be used as an iterator.